### ISSUE TITLE
ESO downloads: no 'verbose' option

### ISSUE THREAD

**Issue by keflavich** [astropy/astroquery#1357]
Some of the data retrieval options in the ESO module lack a `verbose` flag and are excessively verbose in at least one reported case (see astropy mailing list).  We need to make sure that all output can be silenced.

**Comment by kakirastern** [comment]
Will add this when I add some additional keywords to the ESO module as outlined in #1233...

**Comment by kakirastern** [comment]
@keflavich Are these log messages from Requests?

**Comment by keflavich** [comment]
I don't think so.  `requests` doesn't generate any output that I'm aware of

**Comment by bsipocz** [comment]
Actually those "Downloading URL..." lines are coming from the progress bar in `query.py`, so a way to easier silence it (and/or docs how to do it) is what we need. 

Otherwise it's worth the time either to invest in setting up an IDE, or build up command line tools that help in searching in the codebase rather than guessing where certain lines are coming from.

**Comment by kakirastern** [comment]
Yeah, now I can find it too without the underscores as indicated in the astropy email list, at [https://github.com/astropy/astroquery/blob/master/astroquery/query.py#L349](https://github.com/astropy/astroquery/blob/master/astroquery/query.py#L349)... BTW, I wanted to do some cross referencing and couldn't find the PR dealing with this and related verbosity issues as mentioned on the Slack channel. Actually which one is it? (I had tried to find it but couldn't, maybe I overlooked.) It might not be a bad idea to have a note of that PR here too in case of duplicate efforts... Normally it would show up somewhere in this issue if someone is working on it.

**Comment by kakirastern** [comment]
Copying from original emails for reference by contributors...

Inquiry from user:
> I am using astroquery v. 0.3.9 running with python 3.6.
>
> When using eso.retieve_data I get a flood of identical (and hence
> meaningless) download messages of the form:
>
> ...
> Downloading URL
> http://dataportal.eso.org/dataPortal/requests/user_name/401459/SAF/KMOS.2017-02-24T00:52:51.138/KMOS.2017-02-24T00:52
> ...
> Downloading URL
> http://dataportal.eso.org/dataPortal/requests/user_name/401459/SAF/KMOS.2017-02-24T00:52:51.138/KMOS.2017-02-24T00:52
> Downloading URL
> http://dataportal.eso.org/dataPortal/requests/user_name/401459/SAF/KMOS.2017-02-24T00:52:51.138/KMOS.2017-02-24T00:52
> Downloading URL
> http://dataportal.eso.org/dataPortal/requests/user_name/401459/SAF/KMOS.2017-02-24T00:52:51.138/KMOS.2017-02-24T00:52
> ...
> repeated several 100?s of times.
> Is there any way to turn this output off, or at least reduce its quantity?

Response from repo maintainer:
>This looks like an error where the progressbar_or_spinner class is producing a newline each time instead of overwriting the existing line; I suspect this has to do with your terminal settings but I'm not exactly sure how to fix it.  We generally have a `verbose` boolean option that can be set to `False`, but it looks like we omitted it in this case; I consider that a bug so we'll note it and try to fix it.

-- Dated circa February 13, 2019
(Edited for clarity and brevity.)

**Comment by Knusper** [comment]
This error still exists.

**Comment by bsipocz** [comment]
@Knusper - If an issue is open, it's likely it's still relevant. PRs are welcome.

**Comment by Knusper** [comment]
It seems the issue is in astropy.utils where the spinner is drawn. I will
investigate as I didn't see an issue (neither open or closed) in their
bugtracker.

**Comment by nandita11** [comment]
Can I take this up @bsipocz ?

**Comment by bsipocz** [comment]
@nandita11 - yes, absolutely. However, I would advise to wait until the maintainers discuss whether the verbose kwarg or switching to logging is the more preferable.


@keflavich @ceb8 - what do you think, should we still at verbose kwargs, or rather use the logger? @andamian good arguments was pushing my thinking towards the logger only setup rather than verbose printing

**Comment by Knusper** [comment]
Sorry, I think there is a misunderstanding here. This issue is not about turning verbose off and on (this was fixed). This is about an issue with the spinner spamming the output on some systems.

**Comment by ceb8** [comment]
@Knusper That was my bad, I didn't read it closely enough.

@nandita11 Given the above clarification this isn't actually a novice ticket.

**Comment by Knusper** [comment]
I think the bug title also should be changed.  

Title: Downloads from ESO archive spam shell.  Description: see https://github.com/astropy/astroquery/issues/1357#issuecomment-483878012

The relevant code that spams the output is around here: https://github.com/astropy/astroquery/blob/e7ea01c520660251c18b5fefe10f4054b4ced64c/astroquery/query.py#L374

There appears also be a related issue opened by @bsipocz in the astropy bugtracker... So I really have no idea why this happens, in principle ProgressBarOrSpinner should not spam the output like this...  Unfortunately I don't have the time to debug this further.

**Comment by Knusper** [comment]
Ah, and by the way - I was wrong about the verbose switch being fixed.  So this bug here is actually two bugs - if verbose is turned on (cannot be turned off at the moment), it should not be excessively and meaninglessly verbose...

**Comment by Knusper** [comment]
OK - after hunting I am really sure the excessive spamming of the output comes from astropy https://github.com/astropy/astroquery/issues/1357#issuecomment-483878012 

The requested verbose flag here should be easy to implement.  I will try to file a PR soon.

**Comment by pllim** [comment]
I don't think `astroquery` should be using `astropy` progress bar. It is a long-standing bug and I don't see it getting fixed anytime soon. Better to replace it with something else more modern. Also see astropy/astropy#4738

**Comment by Knusper** [comment]
For now I have monkey patched astroquery for my project to avoid the excessive verbosity:
```diff

modified   astroquery/query.py
@@ -411,8 +411,8 @@ class BaseQuery:
             progress_stream = io.StringIO()
 
         with ProgressBarOrSpinner(
-                length, ('Downloading URL {0} to {1} ...'
-                         .format(url, local_filepath)),
+                length,
+                f"Downloading {url} ...",
                 file=progress_stream) as pb:
             with open(local_filepath, open_mode) as f:
                 for block in response.iter_content(blocksize):

```

**Comment by keflavich** [comment]
Yes, `tqdm` would be fine.  

Your monkeypatch just removes the filename reporting, though - is the issue not with the progressbar?

**Comment by Knusper** [comment]
No, from what I understand, the progress bar is fine. But for downloads
from the ESO archive no file size is known - so here `ProgressBarOrSpinner`
will yield a  spinner.

 As I wrote in my bug report in astropy [1]
If the line is too long for the terminal, which is the case when the
filename and the URL is displayed, then every spinner update will create a
new line.
Thus I just made the line shorter for now 🤠


[1]: https://github.com/astropy/astropy/issues/11556
